usr/src/uts/common/ktli/t_koptmgmt.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License, Version 1.0 only
+ * (the "License").  You may not use this file except in compliance
+ * with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
+/*	  All Rights Reserved  	*/
+
+/*
+ * Copyright 2005 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+/*
+ * Copyright 2018 Nexenta Systems, Inc.  All rights reserved.
+ */
+
+/*
+ * kTLI variant of t_optmgmt(3NSL)
+ * Returns 0 on success or an errno value.
+ * Similar to libnsl t_optmgmt.c
+ *
+ * Note: This expects the caller's struct t_optmgmt to contain the
+ * XTI version of struct T_opthdr (used with T_OPTMGMT_REQ == 27)
+ * not the old "struct opthdr" (used with T_SVR4_OPTMGMT_REQ == 9)
+ */
+
+#include <sys/param.h>
+#include <sys/types.h>
+#include <sys/proc.h>
+#include <sys/file.h>
+#include <sys/user.h>
+#include <sys/vnode.h>
+#include <sys/errno.h>
+#include <sys/stream.h>
+#include <sys/ioctl.h>
+#include <sys/stropts.h>
+#include <sys/strsubr.h>
+#define	_SUN_TPI_VERSION 2
+#include <sys/tihdr.h>
+#include <sys/timod.h>
+#include <sys/tiuser.h>
+#include <sys/t_kuser.h>
+#include <sys/kmem.h>
+
+int
+t_koptmgmt(TIUSER *tiptr, struct t_optmgmt *req, struct t_optmgmt *ret)
+{
+	struct strioctl		strioc;
+	struct T_optmgmt_req	*opt_req;
+	struct T_optmgmt_ack	*opt_ack;
+	file_t			*fp;
+	vnode_t 		*vp;
+	char			*ctlbuf = NULL;
+	char			*opt_data;
+	t_scalar_t		optlen;
+	int			ctlsize;
+	int 			retval;
+	int			error;
+
+	fp = tiptr->fp;
+	vp = fp->f_vnode;
+
+	optlen = req->opt.len;
+	if (optlen > 0) {
+		if (req->opt.buf == NULL)
+			return (EINVAL);
+		if (optlen < (t_scalar_t)sizeof (struct T_opthdr)) {
+			/* option buffer should atleast have an t_opthdr */
+			return (EINVAL);
+		}
+		/* sanity limit */
+		if (optlen > 4096) {
+			return (EINVAL);
+		}
+	}
+
+	ctlsize = sizeof (*opt_req) + optlen;
+	ctlbuf = kmem_alloc(ctlsize, KM_SLEEP);
+
+	/* LINTED E_BAD_PTR_CAST_ALIGN */
```

Comment:
I actually prefer to cast through (void *) for this and avoid the LINTED complaints.  That tends to satisfy other compiler warning systems too.   But I'm also ok with this for now.

Reply:
Just copied from libnsl...